pdns/dnsdistdist/dnsdist-backend.cc

```diff
@@ -1087,14 +1066,10 @@ void ServerPool::addServer(shared_ptr<DownstreamState>& server)
 
 void ServerPool::removeServer(shared_ptr<DownstreamState>& server)
 {
-  auto servers = d_servers.write_lock();
-  /* we can't update the content of the shared pointer directly even when holding the lock,
-     as other threads might hold a copy. We can however update the pointer as long as we hold the lock. */
-  auto newServers = std::make_shared<ServerPolicy::NumberedServerVector>(*(*servers));
   size_t idx = 1;
   bool found = false;
   bool tcpOnly = true;
-  for (auto it = newServers->begin(); it != newServers->end();) {
+  for (auto it = d_servers.begin(); it != d_servers.end();) {
```

Comment:
is there a reason not to use

```cpp
for (auto it : d_servers) {...}
```

Reply:
We erase one entry from the container, which invalidates the iterator, so we need to update it.